### ISSUE TITLE
Convert Java Iterator classes to implement IEnumerable<T>

### ISSUE THREAD

**Issue by clambertus** [apache/lucenenet#279]
<p>The Iterator pattern in Java is equivalent to IEnumerable in .NET.  Classes that were directly ported in Java using the Iterator pattern, cannot be used with Linq or foreach blocks in .NET.</p>

<p><tt>Next()</tt> would be equivalent to .NET's <tt>MoveNext()</tt>, and in the below case, <tt>Term()</tt> would be as .NET's <tt>Current</tt> property.  In cases as below, it will require <tt>TermEnum</tt> to become an abstract class with <tt>Term</tt> and <tt>DocFreq</tt> properties, which would be returned from another class or method that implemented <tt>IEnumerable<TermEnum></tt>.</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre> 
	public abstract class TermEnum : IDisposable
	{
		public abstract bool Next();
		public abstract Term Term();
		public abstract int DocFreq();
		public abstract void  Close();
	public abstract void Dispose();
	}
</pre>
</div></div> 

<p>would instead look something like:</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre> 
	public class TermFreq
	{
		public abstract Term { get; }
		public abstract int { get; }
	}

public abstract class TermEnum : IEnumerable<TermFreq>, IDisposable
{
// ...
}
</pre>
</div></div>

<p>Keep in mind that it is important that if the class being converted implements <tt>IDisposable</tt>, the class that is enumerating the terms (in this case <tt>TermEnum</tt>) should inherit from both <tt>IEnumerable<T></tt> <b>and</b> <tt>IDisposable</tt>.  This won't be any change to the user, as the compiler automatically calls <tt>IDisposable</tt> when used in a <tt>foreach</tt> loop.</p>
<i>JIRA link - <a href="https://issues.apache.org/jira/browse/LUCENENET-469">[LUCENENET-469]</a> created by ccurrens</i>

**Comment by NightOwl888** [comment]
This was partially addressed in #341. However, there are still a few iterators to convert, if sensible.

1. `DocIdBitSetIterator` and all subclasses
2. `CharacterIterator`
3. `BreakIterator` and all subclasses

**Comment by paulirwin** [comment]
I looked into the remaining types a little bit, and none of them map cleanly to `IEnumerable<T>`, unless we wanted to add an _additional_ `IEnumerable<T>` interface to them instead of trying to replace their Iterator-ness.

## DocIdSetIterator (and all of its subclasses)

Apart from the monumental amount of work here if we were to change this design in a breaking way (with well over 100 derived types), DocIdSetIterator is perhaps the closest of the three remaining types to being able to cleanly map to `IEnumerable<T>`. `NextDoc()` maps pretty cleanly to the `IEnumerator<int>` pattern. I think our best path is going to be adding `IEnumerable<T>` here in a non-breaking way, rather than trying to change the existing implementation. That way, anyone that wants to enumerate over it may do so, but the existing methods are not affected.

A quick implementation on DocIdSetIterator might look like:

```c#
public IEnumerator<int> GetEnumerator() => new DocIdSetEnumerator(this);

IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

private class DocIdSetEnumerator : IEnumerator<int>
{
    private readonly DocIdSetIterator outerInstance;
    private int current;

    public DocIdSetEnumerator(DocIdSetIterator outerInstance)
    {
        this.outerInstance = outerInstance;
        current = -1;
    }

    public int Current => current;

    object IEnumerator.Current => current;

    public void Dispose()
    {
    }

    public bool MoveNext()
    {
        current = outerInstance.NextDoc();

        return current != NO_MORE_DOCS;
    }

    public void Reset()
    {
        throw new System.NotSupportedException();
    }
}
```

This way, all subclasses will automatically inherit this behavior.

If that design sounds good, I can get a PR for that.

## CharArrayIterator 

I believe this type is what was referred to as `CharacterIterator` in @NightOwl888's last post from 2020, but let me know if I'm wrong. This type supports bidirectional iteration, and I think there is a nice symmetry currently with `Next` and `Previous`. I don't think it's a good idea for us to change this to support `IEnumerable<char>`, although again I could see adding an implementation similar to the one above for DocIdSetIterator that just calls `Next` on the CharArrayIterator.

## BreakIterator

I'm assuming this is referring to BreakIterator in ICU4N, in which case that should be an issue there, although similarly to CharArrayIterator, it has a nice bidirectional symmetry currently. I think we should keep its existing design, and just consider adding `IEnumerable<int>` to it in a non-breaking way.

**Comment by NightOwl888** [comment]
## DocIdSetIterator

I tried to convert this a while back and I pretty much came to the conclusion it is better the way that it is. The main issue was the `DocsAndPositionsEnum` which seemed to be able to switch between enumerating docs and enumerating positions (although I found no such switching anywhere in the code, so maybe that is an invalid assumption.

## CharArrayIterator

This technically has no purpose in .NET. They were used in Java to potentially fix bugs in BreakIterators in broken JREs. There is at least 1 other implementation of `CharacterIterator` that is also completely useless in .NET.

## BreakIterator

Technically, `BreakIterator` was an interface in the JDK and the design in ICU4J is an "improved" design that does not subclass it. In .NET, we merged them into one thing because the BCL is missing anything like `BreakIterator`.

In any case, the design needs rework to be compatible with `ReadOnlySpan<char>` because `CharacterIterator` wraps the `chars[]` and `BreakIterator` wraps `CharacterIterator`. My thought is to add a seam so the rules engine can be used without the rest of the design and then it would be possible to create a `ValueBreakIterator` ref struct. But it is tricky because the purpose of the `CharacterIterator` piece is to allow input as `UTF-8` without changing the rest of the design and working out a way to make a more optimized design extensible requires some thought. Ideally, `CharacterIterator` would be eliminated completely.

Anyway, that does mean it is not related to this project.

------------------------------------

However, looking at the commit history, it looks like there was some work on this earlier this year (May 7) that is still on one of my branches. I don't know whether that work has been merged yet. This will need to be reviewed before we can close this.

**Comment by paulirwin** [comment]
Also, the `IBytesRefIterator` and `BytesRefIterator` types (and the implemented methods like Next on types like TermsEnum) are marked as Obsolete for removal in the 4.8.0 release candidate. If we still intend to remove those, we should remove them before closing this issue out.

**Comment by paulirwin** [comment]
> However, looking at the commit history, it looks like there was some work on this earlier this year (May 7) that is still on one of my branches. I don't know whether that work has been merged yet. This will need to be reviewed before we can close this.

@NightOwl888 do you mind taking this one to close it out? I could split out the obsolete removals into a separate ticket. It would be awesome to close out this issue after (*checks notes*) 14 years 😆

**Comment by NightOwl888** [comment]
`AbstractAppendingInt64Buffer` has a public iterator that could be converted.

**Comment by paulirwin** [comment]
@NightOwl888  is this the branch you were referring to that you said had some unmerged work? https://github.com/NightOwl888/lucenenet/tree/feature/enumerators-icharacterenumerator

if so, I could look into picking this back up and determining where it left off, and finish it up.

**Comment by paulirwin** [comment]
I'm going to pick this up where @NightOwl888 left off on his branch. Will check these off as I PR them. I'll cherry pick @NightOwl888's commits to retain authorship.

Already done on his branch/master:
- [x] MergedIterator -> MergedEnumerator
- [x] FilterIterator -> FilterEnumerator

WIP on his branch/master:
- [x] ~~ICharacterEnumerator + CharacterEnumeratorWrapper/CharacterIteratorWrapper~~ (split as #1398 for future)
- [x] Remove all obsolete iterator types/methods in RC (split out as its own issue for 4.8.0 RC: #1397)
- [x] ~~BytesRefFSTEnum~~ (is not an iterator in Java)
- [x] ~~Int32sRefFSTEnum~~ (is not an iterator in Java)

Still to do:
- [x] AbstractAppendingInt64Buffer.Iterator (#1396)
- [x] ~~BlockPackedReaderIterator~~ (doesn't fit pattern, see comment below)
- [x] ~~PackedInts.ReaderIterator~~ (doesn't fit pattern, see comment below)

To skip:
- CharArrayIterator (candidate for eventual removal)
- DocIdSetIterator (not worth it, per discussion above)
- BreakIterator (out of scope; external)

**Comment by paulirwin** [comment]
After reviewing BlockPackedReaderIterator and PackedInts.ReaderIterator, I don't think we should do these either. They have a `Next(int count)` overload that doesn't match `IEnumerator<T>`. I think this is the same as the logic of why we feel DocIdSetIterator is not in scope.

Also, the MergedEnumerator and FilterEnumerator work was already in master, so that was already completed.

The FSTEnum types are not iterators in Java, so I do not think we should convert those here.

So, after PR #1396 that I just opened, that just leaves ICharacterEnumerator/CharacterIterator work. I will analyze that and update here.

**Comment by NightOwl888** [comment]
> So, after PR [#1396](https://github.com/apache/lucenenet/pull/1396) that I just opened, that just leaves ICharacterEnumerator/CharacterIterator work. I will analyze that and update here.

@paulirwin - there was an update to the status of this one here: https://github.com/NightOwl888/ICU4N/issues/23#issuecomment-2366826611

The `CharacterIterator` does 3 things for Java users:

1. Allows existing `BreakIterator` implementations to change their input encoding without any other changes.
2. Allows patches for JDK implementations that have broken `CharacterIterator`s.
3. Allows Lucene users to extend `CharacterIterator` for other purposes.

Number 2 doesn't seem like a huge concern for .NET. The 3rd is sort of an open question. Given the fact that `CharacterEnumerator` is extended in several places in Lucene, it isn't clear whether users benefit from this extension point or if it only exists to work around JDK bugs.

That said, the fundamental flaw with carrying this `BreakIterator` and `CharacterIterator` design over to .NET is the fact that once you wrap a sequence of characters or bytes with an interface, you lose the ability to perform the operation on the stack. Microsoft ended up just [re-implementing `BreakIterator` segmentation](https://github.com/dotnet/runtime/blob/v10.0.9/src/libraries/System.Private.CoreLib/src/System/Text/Unicode/TextSegmentationUtility.cs) so they could avoid the heap allocation. I haven't fully analyzed that implementation to determine how feasible it would be to make a pluggable `BreakIterator` rules engine fit with that type of design. The real crux of the issue sinks in when you consider how ICU4C supports both UTF-16 and UTF-8 inputs and the fact that Microsoft simply makes these separate inputs into separate static APIs, making it harder to plug a common rules engine into both of them. Unfortunately, there isn't an existing seam in `RuleBasedBreakIterator` that we can take advantage of like was done with `NumberFormatRules` to pull the rules engine into a separate pluggable concept, and it is complicated by the fact that in .NET that `UCultureInfo` contains the data for a given locale rather than being a key to lookup the data as was done in ICU4J. 

The current thinking is to continue supporting the `BreakIterator`/`CharacterIterator` in ICU4N while experimenting with the above type of approach to lift segmentation behavior up to the stack. That gives Lucene.NET a window to figure out how to migrate that doesn't necessarily need to be done for the 4.8.0 release. But although I have tried [an experimental `ICharacterEnumerator` implementation](https://github.com/NightOwl888/J2N/blob/v2.1.0/src/J2N/Text/StringCharacterEnumerator.cs) already, it seems like converting it without also changing the design of `BreakIterator` is causing more problems than it solves. The only real "problem" with it is that it exposes the `ICU4N.Support` namespace publicly, but that is simple enough to change for the ICU4N release by moving it into the `ICU4N.Text` namespace and setting it up to be marked as "currently the only way available, but we know this isn't how we want it done" type of API status. 

Note that `CharacterIterator` came from the JDK, so it seemed at first like putting it in `J2N` would be the right choice. But given the facts that it works in conjunction with a `BreakIterator` implementation in Lucene (thus requiring a reference to ICU4N anyway) and it actively works against the goal of moving to System.Memory, it seems better to try to phase it out than to support it.

The fact that Lucene.NET extends the `CharacterIterator` interface is just another issue to deal with. Note that for the time being, [ICU4N wraps a `ReadOnlyMemory<char>` instance](https://github.com/NightOwl888/ICU4N/blob/v60.1.0-alpha.440/src/ICU4N/Impl/CSCharacterIterator.cs) for use with `BreakIterator`. Since the only known purpose of these classes in Lucene is to fix JDK bugs that don't exist in .NET, I wonder if it would be best to do away with any public notion of `CharacterIterator` in Lucene.NET and simply expose APIs that accept `string`, `char[]` and `ArraySegment<char>`. We could expose `ReadOnlyMemory<char>` also, but it is a bit more dangerous because it doesn't hold on to a pointer. I haven't yet looked at how feasible that change is. That would potentially prevent Lucene.NET users from attempting to use `CharacterIterator` as a way to extend custom highlighter implementations, giving us the ability to change the design later without breaking public APIs.

It also makes me wonder whether we should marry a stack-based `BreakIterator` (or possibly enumerator) design with stack-based highlighters in the `lucenenet-extensions` repo to experiment with how to redesign the highlighters for use with System.Memory types while still providing the ability to extend them.

**Comment by paulirwin** [comment]
@NightOwl888 Thanks for the detailed write-up. I started investigating this and came to the same conclusions you did (but with far less context/detail). There isn't really a way to do this currently since it's so object-oriented.

I think we can cut this one from scope and revisit it later. It's not an urgent need, anyways. I don't see anyone providing benchmarks and complaining it's slow. We can save it for post-4.8.0. I split that out as #1398 based on your comment.

With that, I think PR #1396 closes out this issue sufficiently for go-live. Please correct me if I'm wrong (or file a new issue).

**Comment by NightOwl888** [comment]
Actually, perhaps we should make an attempt to actively avoid using `CharacterIterator` in Lucene.NET and call one of the other overloads of `BreakIterator.SetText()` (or add more if needed). That way we don't paint our users into a corner thinking that they can extend it. It will certainly make transitioning to a new System.Memory-centric design more straightforward if we disallow it.
